tests/literal.json

```diff
@@ -78,6 +78,10 @@
                 "expression": "`9`",
                 "result": 9
             },
+            {
+              "expression": "`42 `",
+              "result": 42
+            },
```

Comment:
I think this is redundant with the "Literal with trailing whitespace" case in the "Literals" cases, and would recommend instead extending that section to include cases like
```json
        {
          "comment": "Literal with all valid leading and trailing JSON whitespace",
          "expression": "` \t\n\r{\"foo\": true} \t\n\r`",
          "result": {"foo": true}
        },
```
and a selection of non-JSON whitespace like
```json
        {
          "comment": "Literal with non-JSON whitespace U+000C FORM FEED",
          "expression": "`0\f`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+0085 NEXT LINE",
          "expression": "`0\u0085`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+00A0 NO-BREAK SPACE",
          "expression": "`0\u00A0`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+1680 OGHAM SPACE MARK",
          "expression": "`0\u1680`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+2028 LINE SEPARATOR",
          "expression": "`0\u2028`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+3000 IDEOGRAPHIC SPACE",
          "expression": "`0\u3000`",
          "error": "syntax"
        },
```

Reply:
Thanks for the suggestion 👍.

This raises a point with regards to the Python implementation – and possibly the JavaScript implementation as well.
As JEP-12 has been introduced, those compliance tests effectively state that those constructs are now not standards-compliant. However, the Python implementation still supports the _legacy_ syntax and converts them to strings.

```patch
{
      "comment": "Literal with non-JSON whitespace U+1680 OGHAM SPACE MARK",
      "expression": "`0\u1680`",
-      "error": "syntax"
+     "result": "0\u1680"
},
```

I’m currently not too keen to break backwards compatibility as I’m trying hard to [get JMESPath Community adopted](https://github.com/aws/aws-cli/issues/7396) by [various stakeholders](https://github.com/Azure/azure-cli/issues/24542). What do you think?

Reply:
It looks like that issue was [anticipated in JEP-12](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-012-raw-string-literals.md#impact) (emphasis mine):
> Implementations MAY choose to support the old syntax of allowing elided quotes in JSON literal expressions. If an implementation chooses this approach, the implementation SHOULD raise some kind of warning to the user to let them know of the deprecation and possible incompatibility with other JMESPath implementations.
> 
> In order to support this type of variance in JMESPath implementations, all of the JSON literal compliance test cases that involve elided quotes MUST be removed, and **test cases regarding failing on invalid unquoted JSON values MUST not be allowed in the compliance test unless placed in a JEP 12 specific test suite**, allowing implementations that support elided quotes in JSON literals to filter out the JEP 12 specific test cases.

The JEP-12 test suite was apparently never added, but now is the time—it should contain the above non-JSON whitespace cases.

> I’m currently not too keen to break backwards compatibility as I’m trying hard to [get JMESPath Community adopted](https://github.com/aws/aws-cli/issues/7396) by [various stakeholders](https://github.com/Azure/azure-cli/issues/24542). What do you think?

In that case, [jmespath.spec](https://github.com/jmespath-community/jmespath.spec) should be updated to document the mode switch with an optional (or required but associated with runtime syntax errors) `json-value =/ *( json-unescaped / json-escaped )` grammar production[^1] and a section explaining the consequences of an implementation decision regarding if or when to support it. But note also that implementation support for the legacy behavior isn't even internally consistent (e.g., https://github.com/jmespath/jmespath.js/issues/90 ).

[^1]: possibly abstracting `*( json-unescaped / json-escaped )` as something like `json-string-contents`

Reply:
I noticed that the James already [split compliance tests](https://github.com/jmespath-community/jmespath.py/tree/develop/tests/legacy) with this deprecation in mind.

So I can actually add those tests in the compliant file.

I also went ahead and [fully deprecated](https://github.com/jmespath-community/jmespath.py/blob/e79fdfe33bf0de9a193e740ed9a978c2e061b102/jmespath/lexer.py#L206-L215) pre-JEP12 JSON string literals while [adding a configuration flag](https://github.com/jmespath-community/jmespath.py/blob/e79fdfe33bf0de9a193e740ed9a978c2e061b102/jmespath/visitor.py#L77-L81) to maintain backwards compatibility.